### PR TITLE
Ion Serde for partiql values.

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- [Closes/Related To] Issue #XXX

## Description
- Ready after PR #1091 is merge. 
- Add Serde between PartiQL value and Ion Value. 
- A short write up explain the serde behavior. 

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  - Yes. 

- Any backward-incompatible changes? **[YES/NO]**
  -  No. 

- Any new external dependencies? **[YES/NO]**
  -  No

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**
Yes. 

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueIonWriter.kt [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import org.partiql.value.StructValue
+import org.partiql.value.SymbolValue
+import org.partiql.value.TextValue
+import org.partiql.value.TimeValue
+import org.partiql.value.TimestampValue
+import java.io.OutputStream
+
+internal class PartiQLValueIonWriter(
+    private val ionWriter: IonWriter,
+) : PartiQLValueWriter {
```
Comment (yliuuuu) [inline partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueIonWriter.kt:62]: I don't think we need to ever failed on writing to Ion. `$bag::[]` is a valid generic ion value, it is during reading we need to concern.

File: partiql-types/src/main/kotlin/org/partiql/value/PartiQLValueReader.kt [hunk 2]
```diff
@@ -1,10 +1,8 @@
...
 
-import java.io.InputStream
-
 public interface PartiQLValueReader : AutoCloseable {
 
-    public fun readValue(input: InputStream): PartiQLValue
+    public fun readValue(): PartiQLValue
 
-    public fun readValues(input: InputStream): Iterator<PartiQLValue>
+    public fun readValues(): Iterator<PartiQLValue>
```
Comment (yliuuuu) [inline partiql-types/src/main/kotlin/org/partiql/value/PartiQLValueReader.kt:7]: Move input stream to builder.

File: partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueIonReader.kt [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        MISSING_ANNOTATION("\$missing"),
+        BAG_ANNOTATION("\$bag"),
+        DATE_ANNOTATION("\$date"),
+        TIME_ANNOTATION("\$time"),
+        GRAPH_ANNOTATION("\$graph")
+    }
+
+    override fun readValue(): PartiQLValue {
+        val value = if (ionReader.next() != null) {
+            readSingleValue(ionReader)
```
Comment (yliuuuu) [inline partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueIonReader.kt:48]: we can potentially move the next() call to reader interface.

File: docs/wiki/serde/PartiQLValue <> ION.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+| NULLABLE_BAG       | LIST                | null -> $bag::null.list else same as BAG                                                                                                                                                                                                                                            | 
+| NULLABLE_LIST      | LIST                | null -> null.list else same as LIST                                                                                                                                                                                                                                                 | 
+| NULLABLE_SEXP      | SEXP                | null -> null.sexp else same as sexp                                                                                                                                                                                                                                                 |
```
Comment (RCHowell) [inline docs/wiki/serde/PartiQLValue <> ION.md:60]:
```suggestion
| NULLABLE_SEXP      | SEXP                | null -> null.sexp else same as SEXP                                                                                                                                                                                                                                                 |
```

File: docs/wiki/serde/PartiQLValue <> ION.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+| NULLABLE_STRUCT    | STRUCT              | null -> null.struct else same as STRUCT                                                                                                                                                                                                                                             |
+
+
+## Read From Ion Data
+PartiQL reserves a couple keywords that is used as ion annotation to extend ION Types. 
+To distinguish between generic ion data and PartiQL specific Ion Data, the reader has two mode: `IonGeneric` and `IonForPartiQL`.
```
Comment (RCHowell) [inline docs/wiki/serde/PartiQLValue <> ION.md:66]:
```suggestion
PartiQL reserves somekeywords which are used as Ion annotations to extend the Ion types. 
To distinguish between generic Ion data and PartiQL data encoded as Ion, the reader has two modes: `IonGeneric` and `IonForPartiQL`. 
```

File: docs/wiki/serde/PartiQLValue <> ION.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+### Ion Generic
+If using `IonGeneric` mode: The ion value annotations are always treated as annotations on PartiQL value. 
+
+For example:
+    - `$missing::null` will be treated as nullValue(annotations = ["missing"])
+
+### Ion For PartiQL
+If using `IonForPartiQL` mode: We examine the **last** annotation before convert to PartiQL Value. If the annotation is PartiQL reserved, we validate Semantics and the constructed PartiQL value may be different. 
+
```
Comment (RCHowell) [inline docs/wiki/serde/PartiQLValue <> ION.md:76]:
```suggestion
When using `IonForPartiQL` mode, the **last** annotation is checked to see if this is an encoded PartiQL value. If the annotation is a PartiQL reserved keyword, then we validate and construct the PartiQL value.

```

File: partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueNullableImpl.kt [hunk 7]
```diff
@@ -317,7 +317,7 @@ internal data class NullableBlobValueImpl(
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
         if (javaClass != other?.javaClass) return false
-        other as NullableClobValueImpl
+        other as NullableBlobValueImpl
```
Comment (RCHowell) [inline partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueNullableImpl.kt:320]: Nice catch

File: partiql-types/src/main/kotlin/org/partiql/value/io/PartiQLValueTextWriter.kt [hunk 8]
```diff
@@ -293,22 +293,22 @@ internal class PartiQLValueTextWriter(
...
         override fun visitNullableList(v: NullableListValue<*>, ctx: Format?) = when (v.isNull()) {
             true -> "null"
-            else -> visitList(v.promote(), ctx)
+            false -> visitList(v.promote(), ctx)
         }
 
         override fun visitNullableSexp(v: NullableSexpValue<*>, ctx: Format?) = when (v.isNull()) {
             true -> "null"
-            else -> visitSexp(v.promote(), ctx)
+            false -> visitSexp(v.promote(), ctx)
```
Comment (RCHowell) [inline partiql-types/src/main/kotlin/org/partiql/value/io/PartiQLValueTextWriter.kt:306]: You didn't like the 4-char/4-char alignment 😛

**Review by RCHowell — CHANGES_REQUESTED** [review]

File: partiql-types/src/main/kotlin/org/partiql/value/io/PartiQLValueTextWriter.kt (hunk 8 above)
Comment (yliuuuu) [reply partiql-types/src/main/kotlin/org/partiql/value/io/PartiQLValueTextWriter.kt:306]: I have such a love-hate relationship with Kotlin's exhaustive when statement :)

**Review by RCHowell — APPROVED** [review]

**Comment by RCHowell** [comment]
Thank you for this. We could reduce some of the repetition, but it would not change the behavior. Also, I acknowledge your choice to explicitly list all annotation type cases

**Pull request merged**